knimin/lib/qiita_client.py

```diff
@@ -0,0 +1,186 @@
+
+        # Set up oauth2
+        self._client_id = client_id
+        self._client_secret = client_secret
+        self._authenticate_url = "%s/qiita_db/authenticate/" % self._server_url
+
+        # Fetch the access token
+        self._fetch_token()
+
+    def _fetch_token(self):
+        """Retrieves an access token from the Qiita server
+
+        Raises
+        ------
+        ValueError
+            If the authentication with the Qiita server fails
+        """
+        data = {'client_id': self._client_id,
+                'client_secret': self._client_secret,
+                'grant_type': 'client'}
+        r = requests.post(self._authenticate_url, verify=self._verify,
+                          data=data)
+        if r.status_code != 200:
+            raise ValueError("Can't authenticate with the Qiita server")
+        self._token = r.json()['access_token']
+
+    def _request_oauth2(self, req, url, as_json=False, **kwargs):
+        """Executes a request using OAuth2 authorization
+
+        Parameters
+        ----------
+        req : function
+            The request to execute
+        url : str
+            The url to access in the server
+        as_json : bool, optional
+            If true, encode the request data using JSON
+        kwargs : dict
+            The request kwargs
+
+        Returns
+        -------
+        requests.Response
+            The request response
+        """
+        url = self._server_url + url
+
+        if 'headers' in kwargs:
+            kwargs['headers']['Authorization'] = 'Bearer %s' % self._token
+        else:
+            kwargs['headers'] = {'Authorization': 'Bearer %s' % self._token}
+
+        if 'data' in kwargs and as_json:
+            kwargs['data'] = dumps(kwargs['data'])
+
+        r = req(url, verify=self._verify, **kwargs)
+        r.close()
+        if r.status_code == 400:
+            try:
+                r_json = r.json()
+            except ValueError:
+                r_json = None
+
+            if r_json and r_json.get('error_description') == \
+                    'Oauth2 error: token has timed out':
+                # The token expired - get a new one and re-try the request
+                self._fetch_token()
+                kwargs['headers']['Authorization'] = 'Bearer %s' % self._token
+                r = req(url, verify=self._verify, **kwargs)
+
+        try:
+            r_json = r.json()
+        except Exception:
+            r_json = None
+
+        return r.status_code, r_json
+
+    def get(self, url, **kwargs):
+        """Execute a get request against the Qiita server
+
+        Parameters
+        ----------
+        url : str
+            The url to access in the server
+        kwargs : dict
+            The request kwargs
+
+        Returns
+        -------
+        int, dict
+            The status code and the JSON response from the server
+        """
+        return self._request_oauth2(requests.get, url, **kwargs)
```

Comment:
Just do `self._session = requests.Session()` in the constructor and then do `self._session.get` for GET, etc. I don't understand the duplication concern?